### ISSUE TITLE
[HOLD for payment 2024-06-24] [$250] Desktop app startup is slow

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#42558]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.75-0
**Reproducible in staging?:** n
**Reproducible in production?:** n
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @neil-marcellini 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1716480870229429

## Action Performed:
1. Install the desktop app on Mac
2. Open it
3. Quit/close it completely
4. Open it again

## Expected Result:
The app opens quickly, within a few seconds
## Actual Result:
The app takes over 30 seconds to open, then a bunch of time to load a chat. Let's focus on the slow opening part specifically, since that's unique to the desktop app.
## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/7347b5a4-0e8d-480f-acf5-cfc14a0dead7

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0192a06bcd75528339</li>
        <li>Upwork Job ID: 1793745864691560448</li>
        <li>Last Price Increase: 2024-06-06</li>
<li>Automatic offers: </li>
<ul>
<li>eh2077 | Reviewer | 102692308</li>
<li>dominictb | Contributor | 102692310</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @anmurali</details>

**Comment by MelvinBot** [comment]
This has been labelled "Needs Reproduction". Follow the steps here: https://stackoverflowteams.com/c/expensify/questions/16989

**Comment by goldenbear101** [comment]
Contributor details
Your Expensify account email: <irvinwoluchem@gmail.com >
Upwork Profile Link: <https://www.upwork.com/freelancers/~017e81b8541352d771?mp_source=share>

**Comment by goldenbear101** [comment]
Contributor details
Your Expensify account email: [irvinwoluchem@gmail.com](mailto:irvinwoluchem@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~017e81b8541352d771?mp_source=share

**Comment by goldenbear101** [comment]
I have solved the issue and it is now ready for submition

**Comment by eh2077** [comment]
@goldenbear101 Can you post a proposal for this issue? You can refer our https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md

**Comment by eh2077** [comment]
Waiting on proposals

**Comment by goldenbear101** [comment]
Sorry to say but, I do not have the means (connects) to write a proposal on upwork. I request you allow me to operate through Telegram, you would be able to access the work, and only pay after confirmation.

**Comment by eh2077** [comment]
@goldenbear101 You can just post proposals here and we'll review it, see [propose-a-solution-for-the-job](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job)

**Comment by goldenbear101** [comment]
**Proposal**

The desktop app for Mac takes over 30 seconds to open, significantly longer than expected. This issue is unique to the desktop app and affects user experience.

**Root Cause**
The slow startup likely results from:

Slow initialization of application components
Inefficient resource loading
Network requests made during startup
Proposed Changes
To solve the problem, the following changes are recommended:

1. Optimize Initialization
Streamline the initialization process to load only essential components at startup.

```
async function initializeApp() {
    await loadEssentialComponents();
    // Load non-essential components after startup
    setTimeout(loadNonEssentialComponents, 0);
}
```
2. Asynchronous Resource Loading
Load resources asynchronously to avoid blocking the main thread.

```
async function loadResources() {
    const [image, data] = await Promise.all([
        loadImageAsync('path/to/image.png'),
        fetchDataAsync('path/to/data.json'),
    ]);
}
```
3. Network Request Optimization
Delay non-critical network requests until after the app has loaded.

```
async function initializeApp() {
    await loadEssentialComponents();
    setTimeout(() => {
        fetchNonCriticalData();
    }, 0);
}
```
**Expected Result**
The app should open quickly, within a few seconds.

By implementing these changes, the startup time of the desktop app should be significantly reduced.

**Comment by neil-marcellini** [comment]
@dominictb your proposal sounds pretty promising. Would you please do some benchmarking / testing for us to gather evidence about how much faster it starts up when built for arm? If you don't have an arm mac, please provide steps for me to test.

**Comment by eh2077** [comment]
I'm testing @dominictb 's proposal

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by eh2077** [comment]
I'll update testing results tomorrow

**Comment by flodnv** [comment]
@dominictb this sounds related with https://github.com/Expensify/App/issues/39631#issuecomment-2148278445, right?

**Comment by eh2077** [comment]
I can see significant difference of starting time between `universal` and `x64` app running on M1 MacBook. I think it's common to create distributables for different hardwares, eg `x64`, `arm64` and `universal` type bundles for macOS. So, @dominictb 's [proposal](https://github.com/Expensify/App/issues/42558#issuecomment-2145836433) looks good to me.

The other issue seems share some root cause with this issue. #39631 was created earlier though this issue seems has clearer scope. I'm also fine if we decide to put this issue on hold for the older one. Let's defer it to internal engineering team.

🎀👀🎀 C+ reviewed

**Comment by yuwenmemon** [comment]
Unassigning myself as @neil-marcellini is the true Engineer on this issue.

**Comment by eh2077** [comment]
Not overdue, waiting for @neil-marcellini 's review on https://github.com/Expensify/App/issues/42558#issuecomment-2155182911

**Comment by neil-marcellini** [comment]
> I can see significant difference of starting time between `universal` and `x64` app running on M1 MacBook. I think it's common to create distributables for different hardwares, eg `x64`, `arm64` and `universal` type bundles for macOS. So, @dominictb 's [proposal](https://github.com/Expensify/App/issues/42558#issuecomment-2145836433) looks good to me.

Ok great, I agree! Let's build the desktop app as `universal`.

**Comment by neil-marcellini** [comment]
I merged that PR, but I want to ask @dominictb @eh2077 how does this work for updating the app? If I click the update menu item in app with it change to being a universal app? If not, how can we make it do that?

**Comment by eh2077** [comment]
### Checklist

- [x] [@eh2077] The PR that introduced the bug has been identified. Link to the PR: _**I don't think there's a specific PR responsible for this**_
- [x] [@eh2077] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- [x] [@eh2077] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] [@eh2077] Determine if we should create a regression test for this bug. No

**Comment by eh2077** [comment]
We fixed two issues together here
- https://github.com/Expensify/App/issues/39631
- https://github.com/Expensify/App/issues/42558

@mallenexpensify I saw you have handled the payment for the other issue. Can you also take care of this one please? As for the payment arrangement, whatever you all think is best works for me!

**Comment by mallenexpensify** [comment]
Contributor: @dominictb paid $250 via Upwork
Contributor+: @eh2077 paid $250 via Upwork

@eh2077 , the PR to fix this that you reviewed also fixed another bug. If you invested additional time on the review, because their was an increase of scope, you might be due additional compensation.  Comment if so.
